node/src/types/value_or_chunk.rs

```diff
@@ -0,0 +1,207 @@
+pub enum ValueOrChunk<V> {
+    /// Represents a value.
+    Value {
+        /// Serialized `V` type using `ToBytes` impl.
+        data: Bytes,
+        /// Phantom marker.
+        _phantom: PhantomData<V>,
+    },
+    /// Represents a chunk of data with attached proof.
+    ChunkWithProof(ChunkWithProof),
+}
+
+#[derive(Debug, PartialEq, Error, Eq)]
+/// Error returned when constructing an instance of `ValueOrChunk`.
+pub enum ChunkingError {
+    /// Merkle proof construction error
+    MerkleConstruction(MerkleConstructionError),
+    /// Serialization error
+    SerializationError(bytesrepr::Error),
+}
+
+impl Display for ChunkingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            ChunkingError::MerkleConstruction(error) => f
+                .debug_tuple("MerkleConstructionError")
+                .field(&error)
+                .finish(),
+            ChunkingError::SerializationError(error) => {
+                f.debug_tuple("SerializationError").field(&error).finish()
+            }
+        }
+    }
+}
+
+impl<V> ValueOrChunk<V> {
+    /// Creates an instance of `ValueOrChunk::Value` if data is less than `max_chunk_size`
+    /// or a `ValueOrChunk::ChunkWithProof` if it is. In the latter case it will return only the
+    /// `chunk_index`-th chunk of the value's byte representation.
+    pub fn new(data: V, chunk_index: u64, max_chunk_size: usize) -> Result<Self, ChunkingError>
+    where
+        V: ToBytes,
+    {
+        let bytes = ToBytes::into_bytes(data).map_err(ChunkingError::SerializationError)?;
+        if bytes.len() <= max_chunk_size {
+            Ok(ValueOrChunk::Value {
+                data: bytes.into(),
+                _phantom: PhantomData,
+            })
+        } else {
+            let chunk_with_proof = ChunkWithProof::new(&bytes, chunk_index)
+                .map_err(ChunkingError::MerkleConstruction)?;
+            Ok(ValueOrChunk::ChunkWithProof(chunk_with_proof))
+        }
+    }
+}
+
+/// Represents an enum that can contain either a whole trie or a chunk of it.
+pub type TrieOrChunk = ValueOrChunk<Bytes>;
+
+impl<V> Display for ValueOrChunk<V> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            ValueOrChunk::Value { data, .. } => f.debug_tuple("Value").field(data).finish(),
+            ValueOrChunk::ChunkWithProof(chunk) => f
+                .debug_struct("ChunkWithProof")
+                .field("index", &chunk.proof().index())
+                .field("hash", &chunk.proof().root_hash())
+                .finish(),
+        }
+    }
+}
+
+impl<V> ValueOrChunk<V> {
+    const VALUE_TAG: u8 = 0;
+    const CHUNK_TAG: u8 = 1;
+
+    fn tag(&self) -> u8 {
+        match self {
+            ValueOrChunk::Value { .. } => Self::VALUE_TAG,
+            ValueOrChunk::ChunkWithProof(_) => Self::CHUNK_TAG,
+        }
+    }
+}
+
+impl<V: ToBytes> ToBytes for ValueOrChunk<V> {
```

Comment:
We need bytesrepr roundtrip tests for this type.

Reply:
I've completely removed `ToBytes` and `FromBytes` impls for `ValueOrChunk` since it was unused - [`512e06c` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/512e06c27f01cc8cbfeb2bb986ee33922400581a).

Requests/Responses are serialized using `bincode`:

1. Construct a get response - [code](https://github.com/casper-network/casper-node/blob/dev/node/src/components/contract_runtime.rs#L260)
2. Calls to `Message::new_get_response` call `Item::to_serialized` - [code](https://github.com/casper-network/casper-node/blob/dev/node/src/protocol.rs#L170)
3. `to_serialized` calls `bincode::serialize` - [code](https://github.com/casper-network/casper-node/blob/dev/node/src/components/fetcher.rs#L94)`